### DISCUSSION TITLE
How to synchronize hydrogen with an external MIDI Clock Signal?

### DISCUSSION THREAD

**Discussion by slu74** [hydrogen-music/hydrogen#1598] [Hydrogen on Linux]
Hello,
I bought a Behringer Pro-1 out of interest in synthesizers and have discovered a new hobby for me!
Now I would like to accompany the sequences created on the Pro-1 with drums from Hydrogen 1.1.1 for Linux, which I'm just getting into. According to the manual, the Pro-1 should output an external clock signal via MIDI or USB MIDI. Is it possible to use the external MIDI clock signal to synchronize Hydrogen and how? Unfortunately I have not found a way yet.
In addition, I would have to tell Hydrogen in which note values I have programmed the sequence (e.g. 1/16) in order to make the Basedrum sound on 1/4.
Unfortunately, the Behringer Pro-1 does not have a MIDI clock in and the mini jack socket provided for this does not help me here.
I have not yet found a similar request on this topic.

**Comment by theGreatWhiteShark** [comment]
Hey @slu74 ,

Hydrogen does not support settings its tempo via MIDI clock yet.

What kind of tempo changes you are looking for? Continuous ones or is it sufficient to change it at the beginning of a beat? In the later case you could you the Timeline of Hydrogen.

If you require more versatile tempo changes you can use it with another application, like Muse4 or Ardour, and synchronize the tempo using JACK. But that's a little bit tricky in Hydrogen and you should keep the overall measure of the song constant in order for the sync to work.

In any case you should probably give the latest version on the master branch a try as it contains quite a number of improvements in terms of tempo change handling.

**Reply by worikgh** [reply]
I am interested in this too.

I have been using `src/player` that builds `h2player`  headless.  I would like to sync it with other synthesisers on my system.  

Reading the source I see that Hydrogen can be a Jack "timebase slave".  I am not sure what that means.

I called `preferences-> m_bJackTimebaseEnabled = true;` which should make Hydrogen respect the Jack timebase.

So far I have not had success.  I can control Hydrogen by repositioning the Jack time line, but I cannot get control of the clock.

I have seen this in the `Preferences.h` so I know I will be restricted in the sorts of patterns I can use like this

```
        /**                                                                                                                                                     │········
         * Since Hydrogen uses both fixed pattern lengths and recalculates                                                                                      │········
         * the tick size each time it encounters an alternative tempo, its                                                                                      │········
         * transport is incompatible to the BBT information provided by                                                                                         │········
         * the JACK server. Only if the length of each pattern corresponds                                                                                      │········
         * to the measure of the respective bar in the timebase master                                                                                          │········
         * application, the bar information provided by JACK can be used                                                                                        │········
         * directly to determine chosen pattern. If this, however, is not                                                                                       │········
         * the case - which can quite easily happen - a complete history                                                                                        │········
         * of all measure and tempo changes would be required to correctly                                                                                      │········
         * identify the pattern. Since this is not provided by JACK, one                                                                                        │········
         * has to either assume the measure or tempo to be constant or                                                                                          │········
         * that the user took care of adjusting the lengths properly.                                                                                           │········
         */                                                                                                                                                     │········
```

**Comment by theGreatWhiteShark** [comment]
Hey @worikgh ,

> I have been using src/player that builds h2player headless

Awesome! Good to hear it's still in use.

> Reading the source I see that Hydrogen can be a Jack "timebase slave". I am not sure what that means.

In JACK an application can register itself as timebase master and broadcast tempo (and measure) information to other programs. With 'slave' we try to indicate that Hydrogen is listening to and processing these information.

But you always need another app, like `Muse4` or `Ardour`, to broadcast tempo information. To my knowledge there are no standalone programs doing this and it can't be done in e.g. `QJackCtl` either.

>  I would like to sync it with other synthesisers on my system.

Are those synthesizers supporting OSC too? In the upcoming version of Hydrogen - available on `master` - you can set tempo directly using an OSC command. (In MIDI this is unfortunately not possible since the available parameter range [0, 127] can not be mapped to the supported tempo range in a trivial and accessible way)

> I am interested in this too.

Could you open an issue requesting this feature?

**Comment by worikgh** [comment]
> In JACK an application can register itself as timebase master and broadcast tempo (and measure) information to other programs. With 'slave' we try to indicate that Hydrogen is listening to and processing these information.

Yes.  I know that.  I hope to take advantage of that.

> >  I would like to sync it with other synthesisers on my system.

> Are those synthesizers supporting OSC too? In the upcoming version of Hydrogen - available on master - you can set tempo directly using an OSC command. (In MIDI this is unfortunately not possible since the available parameter range [0, 127] can not be mapped to the supported tempo range in a trivial and accessible way)

That is all under my control.  Currently I am using a mixture of Pure Data, straight audio,  and Yoshimi to produce the audio sinals  and using Jack to route it through LV2 plugins.  

Early days yet, and it is all fragile, but I am pleased with the results.

Currently it is all based around live interaction (a key board, a LPX pad, and voice/guitar).  I want to add the concept of programmed beats into my system.  Cutting a long story short it would be very sweet I believe if I could synch hydrogen to a clock signal that I provide.

I think that capability is in the design of hydrogen already.  And it is being a Jack slave.  

I am looking into OSC.  So far I have not successfully initialised the server, but early days.  

Got to say, hydrogen is looking and sounding very impressive.  My musical colleague is hopeless interfacing with a computer and he got the idea straight away and made good use in just a few minutes.  High praise!

**Reply by worikgh** [reply]
I should add I am using an eight gig Raspberry Pi 4 and a Pi Sound sound card.

I do not have cycles to waste, so my plan is to make the whole ball of wax usable headless

**Reply by theGreatWhiteShark** [reply]
> I think that capability is in the design of hydrogen already. And it is being a Jack slave.

Yes. Hydrogen's audio engine will able to work with such a new feature with almost no changes. But apart from the actual implementation of the MIDI clocking support this also has to be integrated into the GUI. It most probably requires a new button to indicate whether Hydrogen is receiving MIDI clock signals and to activate/deactivate it. We have to find a place/make room for it etc. It's probably not going to happen prior to the 1.2 release for which we will do a beta soon. But it's definitely worth to implementing.

Could you open a new issue with feature request?

> Got to say, hydrogen is looking and sounding very impressive. My musical colleague is hopeless interfacing with a computer and he got the idea straight away and made good use in just a few minutes. High praise!

Thanks!

**Comment by worikgh** [comment]
Ok.

I get it now.

Jack Timebase will not get me what I need.  I want tick by tick control of the clock.

@theGreatWhiteShark thank you for your patient attention.  I created an issue here: https://github.com/hydrogen-music/hydrogen/issues/1644

**Comment by theGreatWhiteShark** [comment]
MIDI clock support was implemented in #1910 and is available in version 2.0
